# BME1063 Homework1 report
**潘修齐 2018522077**

[toc]

## Data Preparation

### Data Overview

The fastq data is generated by the software [dwgsim](https://github.com/nh13/DWGSIM) written by `Nils Homer`. The parameters are set as default except the sequencing error rate (in order to simulate high-quality sequencing data, mimicing real data). 3 pair-end datasets are generated based on human genome `hg38` with different read length, namely 50, 70 and 100 bp respectively.

### data regeneration

To regenerate the data, run pbs file on `PBS/HW1-1_DataSimulation/HW1-1_DataSimulation.pbs`

```shell
qsub PBS/HW1-1_DataSimulation/HW1-1_DataSimulation.pbs

# for personal computer, run as shell script:
# bash PBS/HW1-1_DataSimulation/HW1-1_DataSimulation.pbs
```

### detailed parameters

| parameter | value |
|-----------|-------|
| error rate | 0.0007 (quality score ~31.5) |
| number of reads | 100000 |
| random seed | 1063 (course number) |
| outer distance between two ends for pairs | mean 500, stddev 50|
| mutation rate | 0.1% |
| indel fraction | 0.1 |
| random DNA read fraction | 5% |
| number of Ns in a read | maximum 0 |
| stddev for base quality score | 50 |

### DataQC

To show the quality of the data simulated, fastqc was performed upon the data generated. To regenerate QC process, run pbs script `PBS/HW1-2_FASTQC/HW1-2_FASTQC.pbs`

```shell
qsub PBS/HW1-2_FASTQC/HW1-2_FASTQC.pbs

# for personal computers, run as shell script:
# bash PBS/HW1-2_FASTQC/HW1-2_FASTQC.pbs
```

The QC result can be found on `FinalReport/FASTQC`, all simulated data shows good quality.

## Results

Simulated data is aligned to reference genome `hg38` with bwa and bowtie2. Both softwares use a pre-index strategy to speed up the alignment process. For `bwa`, the index is generated using 10-core, 32 Gb ram on HPC. For `bowtie2`, the index is downloaded from the official website.

### Time summary

Time summary is extracted by a script `script/1-time_summary.sh`

```shell
bash script/1-time_summary.sh 
```

The running time (user time) of two softwares are as follows: 

| dataset | BWA time | Bowtie2 time |
|---------|----------|--------------|
| dataset1 | 1m17s   | 1m21s        |
| dataset2 | 1m9s    | 1m29s        |
| dataset3 | 1m13s   | 1m48s        |

Using code in `script/2-vidualization_analyzation.ipynb`, a visualization can be generated as follows:



## Discussion